### DISCUSSION TITLE
GitHub get-tree API limits and recursivity

### DISCUSSION THREAD

**Discussion by alhaddad-nasry** [community/community#23748] [New to GitHub]
<p>Hello, </p>
<p>As per the get-tree API (GET /repos/:owner/:repo/git/trees/:tree_sha):</p>
<p>“If truncated is true, the number of items in the tree array exceeded our maximum limit. If you need to fetch more items, you can clone the repository and iterate over the Git data locally.”</p>
<p>Another online post mentions a GitHub API limit of 1,000 items per tree (<a href="https://github.com/websharks/wp-kb-articles/issues/76" rel="nofollow noopener">https://github.com/websharks/wp-kb-articles/issues/76</a>)</p>
<p><strong>Is there an exact number of items allowed in a tree? If yes, how many?</strong></p>
<p>An example I tried:</p>
<p>The root directory containing README.md, 4,000 files, and two subdirectories, containing 4,000 files each.</p>
<ul>
<li>/</li>
<li>---- README.md</li>
<li>---- 4,000 files in /</li>
<li>---- /dir1</li>
<li>-------- 4,000 files in /dir1</li>
<li>---- /dir2</li>
<li>-------- 4,000 files in /dir2</li>
</ul>
<p>Calling get-tree API is returning the whole list of 12,003 items, which makes the limit of 1,000 items not possible. The attribute “truncated” is always returned in the response with a value of false.</p>
<p>Another question for this example: <strong>If recursive was passed to the API with a value of 0, shouldn’t the response exclude the 8,000 files from the two subdirectories?</strong> Regardless if recursive is passed with a value of 0 or 1, it is always returning all files in the response.</p>

**Comment by nadiajoyce** [comment, marked as answer]
<p>Hi <a class="mention" href="/alhaddad-nasry">@alhaddad-nasry</a>,</p>
<p>Thanks for being here :slight_smile:</p>
<p>We actually have a limit of 100,000 entries. If the <code>recursive</code> parameter is supplied, we will read and return a maximum of 7 MB worth of entries from <code>git ls-tree</code>.</p>
<p>The limit was bumped from 1,000 entries to 100,000 in 2014. The issue linked to in the OP may have been working on slightly outdated assumptions.</p>
<p>This all said, keep in mind that 100,000 is the current limit, but it could change anytime in the future.</p>
<p>I hope this helps! </p>

**Comment by niclaslindberg** [comment]
<p>Hi!</p>
<p>Wouldn’t it be nice to include if get tree (recursive) had the support of:</p>
<p>*) Max depth. </p>
<p>*) Simple file pattern matching.</p>
<p>BR</p>
<p>Niclas</p>

**Comment by jcagme** [comment]
<p>Is there a limit when creating a new tree? When doing a POST of a new tree with around 1.5K the tree is created successfully but when attempting with a 20K tree then I get a 502. Have been looking for documentation about this but have not been lucky so far.</p>
<p>Thanks in advance!</p>

**Comment by gbidsilva** [comment]
<p><a class="mention" href="/nadiajoyce">@nadiajoyce</a>, Is there a documentation where it specify the max limit for Tree API?<br>
It seems that, limits are not included in the : <a href="https://developer.github.com/v3/git/trees/" rel="nofollow noopener">https://developer.github.com/v3/git/trees/</a> page.</p>
